### PR TITLE
[5.2] Show a warning when there are due tasks

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

Show next run, show list by next run and show a warning when there are due tasks

### Testing Instructions

- Load the scheduled tasks
- Notice the ordeing by title and the missing Next Run field from the list
- Apply this patch
- Open that page again
- Confrim that the Next Run field is there as well as its orderd by that field so the next task to run will be shown first

image

- Go to the database -> #__scheduler_tasks take one task and set the `next_execution` date to someting in the past
image

- Reload the scheduled tasks list and notice the message that comes up
image

- Use the "Test " button
- Notice the message is gone, it will come up of course as soon as there is another due task which is not run

### Actual result BEFORE applying this Pull Request

No way to detect not run tasks or get to know the next tasks to trigger.

### Expected result AFTER applying this Pull Request

1. Warning when there are due tasks
2. Option to show the next time a task should run
3. Default ordeing to show the next task to run first

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: https://help.joomla.org/proxy?keyref=Help51:Scheduled_Tasks&lang=en
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: administrator/components/com_scheduler/src/Model/TasksModel.php [hunk 1]
```diff
@@ -440,7 +440,7 @@ protected function populateState($ordering = 'a.title', $direction = 'ASC'): voi
     }
 
     /**
-     * Check if we have any enabled due tasks and no locked tasks.
+     * Check if we have any enabled due but no locked tasks.
```
Comment (brianteeman) [inline administrator/components/com_scheduler/src/Model/TasksModel.php]: This comment, and subsequent changes, makes no sense. I am sure its just a "lost in translation" issue and I am happy to help but I just dont understand what it is trying to say

Comment (zero-24) [reply administrator/components/com_scheduler/src/Model/TasksModel.php]: Will revert the change. Thanks

```suggestion
     * Check if we have any enabled due tasks and no locked tasks.
```

File: administrator/components/com_scheduler/src/Model/TasksModel.php [hunk 2]
```diff
@@ -468,4 +468,15 @@ public function hasDueTasks(Date $time): bool
         // False if we don't have due tasks, or we have locked tasks
         return $taskDetails && $taskDetails->due_count && !$taskDetails->locked_count;
     }
+
+    /**
+     * Check if we right now now have any enabled due but no locked tasks.
```
Comment (zero-24) [inline administrator/components/com_scheduler/src/Model/TasksModel.php]:
```suggestion
     * Check if we have right now any enabled due tasks and no locked tasks.
```

File: administrator/language/en-GB/com_scheduler.ini [hunk 3]
```diff
@@ -80,10 +80,15 @@ COM_SCHEDULER_LABEL_TIMES_FAIL="Times Failed"
...
 COM_SCHEDULER_LAST_RUN_DATE="Last Run Date"
 COM_SCHEDULER_LAST_RUN_DESC="Last Run descending"
+COM_SCHEDULER_NEXT_RUN_ASC="Next Run ascending"
+COM_SCHEDULER_NEXT_RUN_DATE="Next Run Date"
+COM_SCHEDULER_NEXT_RUN_DESC="Next Run descending"
+COM_SCHEDULER_NEXT_RUN_MANUAL="Manual"
 COM_SCHEDULER_MANAGER_TASKS="Scheduled Tasks"
 COM_SCHEDULER_MANAGER_TASK_EDIT="Edit Task"
 COM_SCHEDULER_MANAGER_TASK_NEW="New Task"
 COM_SCHEDULER_MANAGER_TOOLTIP_TASK_FAILING="Task failed. Exit code: %1$d"
```
Comment (Quy) [inline administrator/language/en-GB/com_scheduler.ini]:
```suggestion
COM_SCHEDULER_MANAGER_TASKS="Scheduled Tasks"
COM_SCHEDULER_MANAGER_TASK_EDIT="Edit Task"
COM_SCHEDULER_MANAGER_TASK_NEW="New Task"
COM_SCHEDULER_MANAGER_TOOLTIP_TASK_FAILING="Task failed. Exit code: %1$d"
COM_SCHEDULER_NEXT_RUN_ASC="Next Run ascending"
COM_SCHEDULER_NEXT_RUN_DATE="Next Run Date"
COM_SCHEDULER_NEXT_RUN_DESC="Next Run descending"
COM_SCHEDULER_NEXT_RUN_MANUAL="Manual"
```

**Comment by Quy** [comment]
In list view, sort indicator does not match dropdown.

**Comment by zero-24** [comment]
Please try again @Quy

File: administrator/language/en-GB/com_scheduler.ini [hunk 4]
```diff
@@ -84,6 +84,11 @@ COM_SCHEDULER_MANAGER_TASKS="Scheduled Tasks"
...
+COM_SCHEDULER_NEXT_RUN_ASC="Next Run ascending"
+COM_SCHEDULER_NEXT_RUN_DATE="Next Run Date"
+COM_SCHEDULER_NEXT_RUN_DESC="Next Run descending"
+COM_SCHEDULER_NEXT_RUN_MANUAL="Manual"
+COM_SCHEDULER_MSG_DUETASKS="There is at least one due task which should have already run. Please make sure that at least one cron scheduler is enabled and running."
 COM_SCHEDULER_MSG_MANAGE_NO_TASK_PLUGINS="There are no task types matching your query."
 COM_SCHEDULER_NEW_TASK="New Task"
```
Comment (Quy) [inline administrator/language/en-GB/com_scheduler.ini]:
```suggestion
COM_SCHEDULER_MSG_DUETASKS="There is at least one due task which should have already run. Please make sure that at least one cron scheduler is enabled and running."
COM_SCHEDULER_MSG_MANAGE_NO_TASK_PLUGINS="There are no task types matching your query."
COM_SCHEDULER_NEW_TASK="New Task"
COM_SCHEDULER_NEXT_RUN_ASC="Next Run ascending"
COM_SCHEDULER_NEXT_RUN_DATE="Next Run Date"
COM_SCHEDULER_NEXT_RUN_DESC="Next Run descending"
COM_SCHEDULER_NEXT_RUN_MANUAL="Manual"
```

Comment (zero-24) [reply administrator/language/en-GB/com_scheduler.ini]: thanks

**Comment by richard67** [comment]
@zero-24 Is this a bug fix or a new feature?
If bug fix, then the bug may also apply to 4.4-dev and so should be fixed there, too (or only there and later be merged up to 5.1-dev). The bug might not be as visible on the 4.4-dev branch as it is here because in 4.4 we did not use the task scheduler for cyclic core stuff, but it might apply also there.
But if new feature: Shouldn't it be made for 5.2-dev?

**Comment by zero-24** [comment]
Moved to 5.2 looking forward to tests on this PR so it can be merged.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on c850fd8a546600d25503f5e51e7ef8b3eb2e8f56<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43491">issues.joomla.org/tracker/joomla-cms/43491</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on c850fd8a546600d25503f5e51e7ef8b3eb2e8f56<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43491">issues.joomla.org/tracker/joomla-cms/43491</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43491">issues.joomla.org/tracker/joomla-cms/43491</a>.</sub>

**Comment by pe7er** [comment]
Thanks @zero-24 !

**Pull request merged**